fix(forms): add dataType and resolved minMax validator issue

iRealNirmal opened the pull request:
Added data type of null and undefined for validators and resolved issue where min and max validator wasn't working when value was

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
Right now angular app isn't working in strict mode when any validator value is set to null also when min and max validator value is set to 0 it's misbheaving

Issue Number: #42267 


## What is the new behavior?
Added data type of null and undefined in validator where it was required and now min and max validator is working for 0 as well

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

iRealNirmal committed:
fix(forms): allow max/min/maxlength/minlength/pattern validators to be bound to `null`

If the validator is bound to `null` then no validation occurs and attribute is not added in DOM.

Closes #42267

iRealNirmal commented:
@sonukapoor @petebacondarwin can you also help for this PR review ?

petebacondarwin commented on packages/forms/src/directives/validators.ts:
```diff
@@ -579,8 +586,10 @@ export class MinLengthValidator implements Validator, OnChanges {
   private _createValidator(): void {
-    this._validator = minLengthValidator(
-        typeof this.minlength === 'number' ? this.minlength : parseInt(this.minlength, 10));
+    if (this.minlength) {
+      this._validator = minLengthValidator(
+          typeof this.minlength === 'number' ? this.minlength : parseInt(this.minlength, 10));
+    }
```
So what you are suggesting here is that if the `minlength` is `0` then there should be no minlength validation. I see that this does indeed bring the native and Angular validation for this value inline.

@AndrewKushnir - what do you think about this? I think we should align native and Angular validation. The question is whether to align it so that `0` causes validation or not.

AndrewKushnir replied:
+1 that we should align with the native validation. Also, I think we should be explicit about the values that trigger validation, i.e. either have a more precise check, such as:
```
const minlength = this.minlength;
if (minlength != null /* not `null` and not `undefined` */ && minlength > 0) {
```
or have a comment:
```
// Enable validation for any truthy value.
// If the value is `0`, we also skip validator initialization since it's a noop.
if (this.minlength) {
```
I'd prefer the first (more explicit) condition.

petebacondarwin commented on packages/forms/src/directives/validators.ts:
```diff
@@ -698,7 +712,9 @@ export const PATTERN_VALIDATOR: any = {
-  host: {'[attr.pattern]': 'pattern ? pattern : null'}
+  host: {
+    '[attr.pattern]': 'pattern ? pattern : null'
+  }  // if value is 0 or null, this attribute will not be added in DOM
```
The current type for the `pattern` property is `string|RegExp`. So the value will never be `0` nor `null`.

I think the `pattern` property below needs its type updating to something similar to the other validators: `string|RegExp|null|undefined`. Then here the comment would be something like:

```ts
// if value is "", null or undefined then this attribute is not added to the DOM
```

petebacondarwin commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -4192,6 +4192,63 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
         });
       });
 
+      it('should remeove attriute min and max if value is null', () => {
```
```suggestion
      it('should remove attribute min and max if value is null', () => {
```

petebacondarwin commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -4192,6 +4192,63 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
+          selector: 'min-max-null',
+          template: `
+            <form [formGroup]="form">
+              <input formControlName="control" [min]="null" [max]="null" randomAttribute="randomValue" [minLength]="null"  [maxLength]="null"/>
```
I don't think you want the `minlength` and `maxlength` bindings in this test.

```suggestion
              <input formControlName="control" [min]="null" [max]="null" randomAttribute="randomValue">
```

Also note that in HTML5 the `input` element is not self-closing.

petebacondarwin commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -4192,6 +4192,63 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
+            </form>
+          `
+        })
+        class MinMaxComponent {
+                    form = new FormGroup({control: new FormControl()});
+                 
+        }
+
+                const fixture = initTest(MinMaxComponent);
+
+                const minAttribute = fixture.debugElement.query(By.css('[min]'));
+                expect(minAttribute).toBeNull();
+
+                const maxAttribute = fixture.debugElement.query(By.css('[max]'));
+                expect(maxAttribute).toBeNull();
+
+                const randomAttribute = fixture.debugElement.query(By.css('[randomAttribute]'));
+                expect(randomAttribute).toBeDefined();
+             
+      });
```
The formatting of this code looks odd.

petebacondarwin commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -4192,6 +4192,63 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
+             
+      });
+
+            it('should remove attriute minLength and maxLength if value is null', () => {
```
```suggestion
            it('should remove attribute minLength and maxLength if value is null', () => {
```

petebacondarwin commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -4192,6 +4192,63 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
+             
+      });
+
+
```
Could you add tests for all these validators when the binding value is `0`?

petebacondarwin replied:
Also can you check the tests for `pattern` cover all cases too?

petebacondarwin commented on packages/forms/test/template_integration_spec.ts:
```diff
@@ -1821,7 +1821,61 @@ import {NgModelCustomComp, NgModelCustomWrapper} from './value_accessor_integrat
          }));
+      it('should not include the min and max validators for null',
+           fakeAsync(() => {
+                      @Component({template: '<input type="input" min="null" max="null">'})
```
I think that this is binding the string `"null"` rather than the value `null`. Should it not be

```suggestion
                      @Component({template: '<input type="input" [min]="null" [max]="null">'})
```

petebacondarwin reviewed (changes requested):
Thanks for putting this together @iRealNirmal. I have a few comments and some things that I would like @AndrewKushnir to confirm.

iRealNirmal commented on packages/forms/src/directives/validators.ts:
```diff
@@ -551,7 +557,8 @@ export class MinLengthValidator implements Validator, OnChanges {
   @Input()
-  minlength!: string|number;  // This input is always defined, since the name matches selector.
+  minlength!: string|number|null|
+      undefined;  // This input is always defined, since the name matches selector.
```
yes seems  I missed removing `!`  mark

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -167,14 +167,14 @@ export const MAX_VALIDATOR: StaticProvider = {
    * Tracks changes to the max bound to this directive.
    */
-  @Input() max!: string|number;
+  @Input() max!: string|number|null|undefined;
```
I'm a bit unsure that we should allow `undefined` to be a possible type. From the template perspective that means that it's ok to use `<input [formControl]="ctrl" min>` or `<input [formControl]="ctrl" max>`, which doesn't look valid. I think the use-case that was described in the linked issue is that it'd be convenient to pass a value that would disable a validator, i.e. something like:

```
// in TS code
class MyComponent {
  myDataModel = {min: null};
}

// in a template
<input [formControl]="ctrl" [min]="myDataModel.min">
```

So I think the input should be required, but also allow `null` (to allow enable/disable the check dynamically).

iRealNirmal replied:
yes I agree as well, Even I wasn't sure to add but I thought you might have other thought and want this type when I read your comment on raised issue. Let me remove it.

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -167,14 +167,14 @@ export const MAX_VALIDATOR: StaticProvider = {
       'input[type=number][max][formControlName],input[type=number][max][formControl],input[type=number][max][ngModel]',
   providers: [MAX_VALIDATOR],
-  host: {'[attr.max]': 'max ? max : null'}
+  host: {'[attr.max]': 'max'}
```
I think this might be a good place for nullish coalescing (`??`), since this is one of its main use-cases: if there is a `null` or `undefined` - user default value (`null`), otherwise - use provided value.
```suggestion
  host: {'[attr.max]': 'max ?? null'}
```
The only value that we are going to cover with this would be `undefined` though, in all other cases `max` and `max ?? null` would be equivalent.

iRealNirmal replied:
yes sounds good

AndrewKushnir reviewed (commented):
Thanks for the changes @iRealNirmal. I've left a few comments.

I think we should actually split the changes into 2 different PRs and land them separately:

1. The fix for the `0` case for `min` and `max` validators (use nullish coalescing in host bindings) and `minlength`/`maxlength` pair too. We can land the fix asap once it's ready and passes al the verification stages (review, CI, tests in Google's codebase).
2. Add the feature of reacting to the `null` value as a signal to disable validation. This change might be a bit more tricky: it involves type changes and we'd need to make sure all validators are consistent and we have sufficient test coverage. It would probably take some time and a few iterations to discuss and test everything + since it's a feature we'd need to land it in the next minor version (12.1.0).

Thank you.

iRealNirmal commented:
@AndrewKushnir regarding native validation check, what I have wrote if we keep then it will check for
null
undefined
NaN
empty string ''
0
false

but still if you want me to do strict checking as you suggested, I will update it.

iRealNirmal commented:
@AndrewKushnir and @petebacondarwin I will drop comment when PR is ready for review again(might be in a day). Meanwhile ignore any changes.

iRealNirmal commented:
Keeping this PR on hold until PR [https://github.com/angular/angular/pull/42412](https://github.com/angular/angular/pull/42412) is merged.

PowerKiKi commented on packages/forms/src/directives/validators.ts:
```diff
@@ -167,14 +167,14 @@ export const MAX_VALIDATOR: StaticProvider = {
    * Tracks changes to the max bound to this directive.
    */
-  @Input() max!: string|number;
+  @Input() max!: string|number|null;
```
I think we should explicitly allow `undefined` in addition to `null`.

```suggestion
  @Input() max: string|number|null|undefined;
```

AndrewKushnir replied:
@PowerKiKi could you please elaborate a little more on why you think `undefined` should be included?

Adding `undefined` would allow for syntax like this: `<input ngModel max>`, i.e. where `min` and `max` attributes would be present without any value and that would not trigger a template type checking error.

iRealNirmal replied:
@PowerKiKi as per conversation in this ticket we are only going to support only null type here, If you have use case of undefined it will be easy to include your suggested change.

Please read conversation here.
[https://github.com/angular/angular/pull/42378#discussion_r641073378](https://github.com/angular/angular/pull/42378#discussion_r641073378)

PowerKiKi replied:
My use-case for undefined was briefly described in #42267. I have a custom component that allows me to configure an `<input type="number">` (amongst other things). Usage for the custom component would be like:

```html
<my-cmp [config]="config"></my-cmp>
```

The component being declared with something like:

```ts
type Config = {min?: number, otherThings?: string};

@Component({
    selector: 'my-cmp',
    template: '<input type="number" [min]="config.min">',
})
class MyCmp {
    @Input() config: Config;
}
```

Because I don't want to force to always specify the entire configuration, some (or all) fields are optional, hence `undefined`-able. But then I cannot just pass them as shown above. It would not compile in strict mode. I would have to jump through a few hoops.

A possibly annoying hoop, albeit common, would be to merge the incoming config, with default values and storing the result as a complete non-undefined config. Probably something like that instead:


```diff
 type Config = {min?: number, otherThings?: string};

 @Component({
     selector: 'my-cmp',
-    template: '<input type="number" [min]="config.min">',
+    template: '<input type="number" [min]="completeConfig.min">',
 })
 class MyCmp {
+    private readonly defaultConfig: Required<Config> = {min: null, otherThings: ''};
+    public completeConfig: Required<Config> = this.defaultConfig;
-    @Input() config: Config;
+    @Input() set config(value: Config) { this.completeConfig = {...this.defaultConfig, ...value}; }
 }
```

Another, simpler, hoop would be to spread default values all over the template like so:

```diff
 type Config = {min?: number, otherThings?: string};

 @Component({
     selector: 'my-cmp',
-    template: '<input type="number" [min]="config.min">',
+    template: '<input type="number" [min]="config.min ?? null">',
 })
 class MyCmp {
     @Input() config: Config;
 }
```

I am not entirely sure which was would have my preference. First one feels more structured and readable, but a bit more boilerplate.

So I guess _there are_ solutions to my use-case even without angular core supporting `undefined`, but I wanted to be sure that this use-case was considered before making a decision.

The other, problematic, use-case of allowing  `<input [formControl]="ctrl" min>`, as described in https://github.com/angular/angular/pull/42378#discussion_r641073378 did not cross my mind until now. And I suppose it would be enough to justify me having to jump the hoops I described. WDTY ?

AndrewKushnir replied:
@PowerKiKi thanks for the reply and describing the use-case. I think that we'd need to make sure that there is an explicit value provided in cases when you want to dynamically disable min/max validator.

Another risk of having `undefined` support might be that it might hide real errors in templates. For example if you have `<input type="number" [min]="config.min">`, the `config.min` field might be missing on the `config` by mistake, thus being `undefined` in that expression and the Forms logic would treat it as "no validator needed here" instead of raising an error.

PowerKiKi replied:
> it might hide real errors in templates

Excellent point ! 👍

 Let's go ahead without support of `undefined`, and I'll deal with it in my own code with one of the workarounds mentioned earlier.

PowerKiKi commented on packages/forms/src/directives/validators.ts:
```diff
@@ -227,14 +227,14 @@ export const MIN_VALIDATOR: StaticProvider = {
    * Tracks changes to the min bound to this directive.
    */
-  @Input() min!: string|number;
+  @Input() min!: string|number|null;
```
Idem

```suggestion
  @Input() min: string|number|null|undefined;
```

iRealNirmal replied:
Same as  above comment, go through

[https://github.com/angular/angular/pull/42378#discussion_r641073378](https://github.com/angular/angular/pull/42378#discussion_r641073378)

iRealNirmal commented:
Due to some rebase issue from my side, noise is created and multiple person got request for code review, please pardon for it.

alan-agius4 commented:
@iRealNirmal, I removed the extra reviewers. Kindly rebase. Thanks.

iRealNirmal commented:
> @iRealNirmal, I removed the extra reviewers. Kindly rebase. Thanks.

@AndrewKushnir @alan-agius4  I think I have to open new PR as base branch was incorrect and If I update it to origin/master I am getting over 200 conflicts even considering I resolve all it might again send request to all contributors.

although if you have  any other suggestion then feel free to share it.

AndrewKushnir commented:
> @AndrewKushnir @alan-agius4 I think I have to open new PR as base branch was incorrect and If I update it to origin/master I am getting over 200 conflicts even considering I resolve all it might again send request to all contributors.

@iRealNirmal I think it'd be great to preserve this PR if possible since it contains some discussion/context. Please let me know if you want me to try rebasing it.

iRealNirmal commented:
> > @AndrewKushnir @alan-agius4 I think I have to open new PR as base branch was incorrect and If I update it to origin/master I am getting over 200 conflicts even considering I resolve all it might again send request to all contributors.
> 
> @iRealNirmal I think it'd be great to preserve this PR if possible since it contains some discussion/context. Please let me know if you want me to try rebasing it.

yes @AndrewKushnir help here will be really appreciated.

iRealNirmal commented:
@AndrewKushnir thanks after some struggle I was able to resolve it.

Also now this PR is in a state where it can be review by you and @petebacondarwin feel free to provide any further input.

petebacondarwin commented on packages/forms/src/directives/validators.ts:
```diff
@@ -579,8 +582,14 @@ export class MinLengthValidator implements Validator, OnChanges {
   private _createValidator(): void {
-    this._validator = minLengthValidator(
-        typeof this.minlength === 'number' ? this.minlength : parseInt(this.minlength, 10));
+    // Enable validation for any truthy value.
+    // If the value is `0`, we reset validator since it's a noop.
```
```suggestion
    // Enable validation only for truthy values.
    // Note that `0` will result in no validation.
```

petebacondarwin commented on packages/forms/src/directives/validators.ts:
```diff
@@ -579,8 +582,14 @@ export class MinLengthValidator implements Validator, OnChanges {
+    if (this.minlength) {
+      this._validator = minLengthValidator(
+          typeof this.minlength === 'number' ? this.minlength : parseInt(this.minlength, 10));
+    } else {  // reset to null in case value is set back to `0`
```
I am not sure this comment is necessary.

```suggestion
    } else {
```

petebacondarwin commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -4229,6 +4229,82 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
         });
       });
 
+      it('should remove attribute min and max if value is null', () => {
```
Are these specs in the right place?
I would have expected them to be under the `describe('validations')` block.

petebacondarwin reviewed (changes requested):
I have made a few suggestions. I didn't comment on every place. If you agree, make sure you apply them consistently across the PR?

iRealNirmal commented:
@AndrewKushnir can you please provide your input again for this PR ?

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -656,8 +686,14 @@ export class MaxLengthValidator implements Validator, OnChanges {
-        typeof this.maxlength === 'number' ? this.maxlength : parseInt(this.maxlength, 10));
+    // Enable validation only for truthy values.
+    // Note that `0` will result in no validation.
+    if (this.maxlength) {
+      this._validator = maxLengthValidator(
+          typeof this.maxlength === 'number' ? this.maxlength : parseInt(this.maxlength, 10));
+    } else {
+      this._validator = nullValidator;
+    }
```
Just a quick comment: I think we can refactor this code a bit:
```suggestion
    const maxlength = this.maxlength;
    this._validator = maxlength ? maxLengthValidator(toInteger(maxlength)) : nullValidator;
```
Notes:
- the `toInteger` function can be created (based on existing code) and reused between minlength and maxlength validators (should probably shave off a few bytes)
- having a separate `const maxlength = this.maxlength` would help save a few bytes too since class fields are not minified when local vars are, so the minified code will look like:
```
    var a = this.maxlength;
    this._validator = a ? c(b(a)) : d;
```
vs the current version:
```
    if (this.maxlength) {
      this._validator = a(
          typeof this.maxlength === 'number' ? this.maxlength : parseInt(this.maxlength, 10));
    } else {
      this._validator = b;
    }
```

AndrewKushnir commented on packages/forms/src/directives/validators.ts:
```diff
@@ -167,14 +167,19 @@ export const MAX_VALIDATOR: StaticProvider = {
+      null;  // This input is always defined (but may be `null`), since the name matches selector.
   /** @internal */
   inputName = 'max';
   /** @internal */
```
The line below:
```
normalizeInput = (input: string): number => parseFloat(input);
```
would have an issue where we'll call `parseFloat` with `null` (if the input is defined as `null`), thus getting `NaN`. Btw, the `normalizeInput` argument type should also reflect the `max` input type (also include `number` and `null`).

AndrewKushnir reviewed (commented):
Hi @iRealNirmal, thanks for addressing the feedback 👍 

I've looked at the changes again and I feel a bit uncomfortable to rely on the truthiness to enable/disable validators. I think it's ok to rely on it (and have sufficient docs and tests coverage), but we should try to minimize it and be more explicit. A good example might be the `MaxLengthValidator` validator where we rely on truthiness of an input value 3 times.
1. In host bindings:
```
  host: {
    '[attr.maxlength]': 'maxlength ? maxlength : null'
  }
```
2. In the `validate` function:
```
  validate(control: AbstractControl): ValidationErrors|null {
    return this.maxlength != null ? this._validator(control) : null;
  }
```
3. ... and in the `_createValidator` function:
```
    // Enable validation only for truthy values.
    // Note that `0` will result in no validation.
    if (this.maxlength) {
      this._validator = maxLengthValidator(
          typeof this.maxlength === 'number' ? this.maxlength : parseInt(this.maxlength, 10));
    } else {
      this._validator = nullValidator;
    }
```
As you can see the checks are slightly different as well and in one case we check against `null` (to make it more confusing we do that with non-strict equal so in fact we check against `undefined` as well). I think it's very easy to miss some cases and make a mistake in the future.

My proposal is to refactor the code and add an `enabled` function that would indicate whether validation logic should be used. We can rely on the truthiness of a value inside that function as a way to tell whether validation is enforced, but in other cases the `enabled` function should be used instead. For example, here is how the above-mentioned code would look like (put all in one block):
```
  host: {
    '[attr.maxlength]': 'enabled() ? maxlength : null'
  }
  // ...
  validate(control: AbstractControl): ValidationErrors|null {
    return this.enabled() ? this._validator(control) : null;
  }
  // ...
    this._validator = this.enabled() ?
      maxLengthValidator(
          typeof this.maxlength === 'number' ? this.maxlength : parseInt(this.maxlength, 10)) :
      nullValidator;
```
So I'd like to propose the following plan:

1. Create a new PR where the `MinLengthValidator` and `MaxLengthValidator` pair would be updated to use `enabled` function (the `refactor` commit) and where the input types would be updated to include `null` as a possible value (in a separate commit).
2. In a separate followup PR update the `AbstractValidatorDirective` to work with the `enabled` function + switch the `MinLengthValidator` and `MaxLengthValidator` pair to work with the `AbstractValidatorDirective`. Note: the `AbstractValidatorDirective` class is used with the `MinValidator` and `MaxValidator` and the plan is to use that abstract directive as a base class for *all* other validators, so this step would be a preparation for switching over other validators.
3. Enable `null` support for the `MinValidator` and `MaxValidator`.
4. Incrementally transition over remaining validators to the `AbstractValidatorDirective` class, adding `null` support as a part of the transition.

Note: we can land the changes as they become available (no rush at all) and given the fact that the changes are well scoped and isolated, community and team members can help with different pieces.

Please let me know if more information is needed or if there are any questions, will be glad to help.

Thank you.

iRealNirmal commented:
> Hi @iRealNirmal, thanks for addressing the feedback 👍
> 
> I've looked at the changes again and I feel a bit uncomfortable to rely on the truthiness to enable/disable validators. I think it's ok to rely on it (and have sufficient docs and tests coverage), but we should try to minimize it and be more explicit. A good example might be the `MaxLengthValidator` validator where we rely on truthiness of an input value 3 times.
> 
> 1. In host bindings:
> 
> ```
>   host: {
>     '[attr.maxlength]': 'maxlength ? maxlength : null'
>   }
> ```
> 
>  1. In the `validate` function:
> 
> ```
>   validate(control: AbstractControl): ValidationErrors|null {
>     return this.maxlength != null ? this._validator(control) : null;
>   }
> ```
> 
>  1. ... and in the `_createValidator` function:
> 
> ```
>     // Enable validation only for truthy values.
>     // Note that `0` will result in no validation.
>     if (this.maxlength) {
>       this._validator = maxLengthValidator(
>           typeof this.maxlength === 'number' ? this.maxlength : parseInt(this.maxlength, 10));
>     } else {
>       this._validator = nullValidator;
>     }
> ```
> 
> As you can see the checks are slightly different as well and in one case we check against `null` (to make it more confusing we do that with non-strict equal so in fact we check against `undefined` as well). I think it's very easy to miss some cases and make a mistake in the future.
> 
> My proposal is to refactor the code and add an `enabled` function that would indicate whether validation logic should be used. We can rely on the truthiness of a value inside that function as a way to tell whether validation is enforced, but in other cases the `enabled` function should be used instead. For example, here is how the above-mentioned code would look like (put all in one block):
> 
> ```
>   host: {
>     '[attr.maxlength]': 'enabled() ? maxlength : null'
>   }
>   // ...
>   validate(control: AbstractControl): ValidationErrors|null {
>     return this.enabled() ? this._validator(control) : null;
>   }
>   // ...
>     this._validator = this.enabled() ?
>       maxLengthValidator(
>           typeof this.maxlength === 'number' ? this.maxlength : parseInt(this.maxlength, 10)) :
>       nullValidator;
> ```
> 
> So I'd like to propose the following plan:
> 
> 1. Create a new PR where the `MinLengthValidator` and `MaxLengthValidator` pair would be updated to use `enabled` function (the `refactor` commit) and where the input types would be updated to include `null` as a possible value (in a separate commit).
> 2. In a separate followup PR update the `AbstractValidatorDirective` to work with the `enabled` function + switch the `MinLengthValidator` and `MaxLengthValidator` pair to work with the `AbstractValidatorDirective`. Note: the `AbstractValidatorDirective` class is used with the `MinValidator` and `MaxValidator` and the plan is to use that abstract directive as a base class for _all_ other validators, so this step would be a preparation for switching over other validators.
> 3. Enable `null` support for the `MinValidator` and `MaxValidator`.
> 4. Incrementally transition over remaining validators to the `AbstractValidatorDirective` class, adding `null` support as a part of the transition.
> 
> Note: we can land the changes as they become available (no rush at all) and given the fact that the changes are well scoped and isolated, community and team members can help with different pieces.
> 
> Please let me know if more information is needed or if there are any questions, will be glad to help.
> 
> Thank you.

@AndrewKushnir really impressive inputs, and I agree with your suggestions  and planning part.

To summarise for this suggestion, so after this PR we have to work on 2 more PR
1. where the MinLengthValidator and MaxLengthValidator pair would be updated to use enabled
2. In a separate followup PR update the AbstractValidatorDirective to work with the enabled function 

and after I  update this PR with your suggested changes we should be good with current PR (Obviously if not any other minor comments)

And yes as you still uncomfortable  with truthiness condition I will update it with below suggestion no  all related areas
`const minlength = this.minlength;
if (minlength != null /* not `null` and not `undefined` */ && minlength > 0) {`

Please let me know my understanding is correct or if am missing anything, once you give input again I will work on changes.

AndrewKushnir commented:
> Please let me know my understanding is correct or if am missing anything, once you give input again I will work on changes.

@iRealNirmal thanks for the reply, here is my proposal (I rephrased the plan a bit): 

1. Close this PR and use it mostly as a a reference while creating other PRs (may be add a link to it in the new PRs). This PR contains a collection of changes that we'd need to merge independently in a sequence (to make changes more isolated and easier to review, document and test).
2. Create a new (separate) PR where the `MinLengthValidator` and `MaxLengthValidator` pair would be updated to have a new `enabled` function and use it in the code. Within the same PR we can update the input types to include `null` as a possible value (for the `MinLengthValidator` and `MaxLengthValidator` pair **only**).
3. In the next (separate) PR update the `AbstractValidatorDirective` to work with the `enabled` function + switch the `MinLengthValidator` and `MaxLengthValidator` pair to work with the `AbstractValidatorDirective`. Note: the `AbstractValidatorDirective` class is used with the `MinValidator` and `MaxValidator` and the plan is to use that abstract directive as a base class for *all* other validators, so this step would be a preparation for switching over other validators.
4. Enable `null` support for the `MinValidator` and `MaxValidator` (also in a separate PR) by leveraging the `enabled` logic support in the `AbstractValidatorDirective` class added within the item above.
5. Incrementally (in separate PRs) transition the remaining validators over to the `AbstractValidatorDirective` class, adding `null` support as a part of the transition.

So the changes can be split into 4+ PRs with a clear goal/scope for each PR. At the end we'll have all validators using the same base class and the enable/disable logic would be implemented consistently across all of them. Other benefits of this approach vs having all the changes in a single (potentially quite large) PR:

- we reduce the overall risk of this change
- it'd be easier to review, document and test individual PRs
- some of the work can be done in parallel (items 2, 3 and 5), so other community and team members can help if needed

Please let me know if more details are needed.

iRealNirmal commented:
@AndrewKushnir sounds good plan, closing this PR as we discussed.
